### PR TITLE
Copy android device logs to normalised scenario name directory when running Cucumber tests

### PR DESCRIPTION
**Pull request by anandbagmar**
When running ATD with cucumber tests, the android device logs were not captured post test execution.

This PR addresses this requirement.

We now first normalise the scenario name (and also account for the scenario outlines) and create a new directory in the reports folder. The android device logs are captured in this directory post test execution.

### CODE REVIEW

**Review by saikrishna321 — APPROVED** [review]

**Review by SrinivasanTarget — APPROVED** [review]
CI is unhappy otherwise good to me.

**Comment by anandbagmar** [comment]
Hi @SrinivasanTarget  & @saikrishna321  - do let me know if there is anything else required for this PR. Thanks.

**Comment by saikrishna321** [comment]
@anandbagmar I see this error in the CI 

```
java.lang.Exception: /Users/runner/work/1/s/target/screenshot does not exist
	at com.appium.utils.FileFilterParser.getFilesFromDirectory(FileFilterParser.java:25)
	at com.appium.utils.FileFilterParser.getScreenShotPaths(FileFilterParser.java:79)
	at com.appium.manager.AppiumParallelMethodTestListener.afterInvocation(AppiumParallelMethodTestListener.java:147)
	at org.testng.internal.invokers.InvokedMethodListenerInvoker.invokeListener(InvokedMethodListenerInvoker.java:60)
	at org.testng.internal.BaseInvoker.runInvokedMethodListeners(BaseInvoker.java:55)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:631)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:173)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:816)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:146)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
java.lang.NullPointerException
	at com.appium.utils.FileFilterParser.getScreenShotPaths(FileFilterParser.java:85)
	at com.appium.manager.AppiumParallelMethodTestListener.afterInvocation(AppiumParallelMethodTestListener.java:147)
	at org.testng.internal.invokers.InvokedMethodListenerInvoker.invokeListener(InvokedMethodListenerInvoker.java:60)
	at org.testng.internal.BaseInvoker.runInvokedMethodListeners(BaseInvoker.java:55)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:631)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:173)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:816)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:146)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Jan 11, 2021 5:05:46 AM com.context.SessionContext remove

```

File: src/main/java/com/appium/manager/ATDRunner.java [hunk 1]
```diff
@@ -117,7 +117,9 @@ private void createSnapshotDirectoryFor() {
             String deviceId = udid.getDevice().getUdid();
             File file = new File(
                     System.getProperty(USER_DIR)
-                            + FileLocations.SCREENSHOTS_DIRECTORY + os + "/"
+                            + FileLocations.SCREENSHOTS_DIRECTORY
+                            + os
+                            + File.pathSeparator
```
Comment (SrinivasanTarget) [inline src/main/java/com/appium/manager/ATDRunner.java:122]: Isn't it `File.separator` instead?

**Comment by anandbagmar** [comment]
ooh. good catch. Fixed it.


Regards,


Anand Bagmar
about.me/anand.bagmar
[image: Anand Bagmar on about.me]
  <http://about.me/anand.bagmar>

**Comment by SrinivasanTarget** [comment]
@anandbagmar some changes here are breaking the windows cucumber support. 

```
WindowsRunCukes>AbstractTestNGCucumberTests.runScenario:31 ï¿½ NullPointer
```

**Comment by anandbagmar** [comment]
Found the source of the npe in windows and fixed it. Windows build is green
now.


Regards,

Anand Bagmar

File: src/main/java/com/appium/manager/AppiumDevice.java [hunk 2]
```diff
@@ -86,20 +86,20 @@ private boolean isNativeAndroid() {
     public String startDataCapture(String scenarioName, Integer scenarioRunCount)
             throws IOException, InterruptedException {
         String fileName = String.format("/run-%s", scenarioRunCount);
-        File logFile = null;
         if (isNativeAndroid()) {
             String udid = this.getDevice().getUdid();
             fileName = String.format("/%s-run-%s", udid, scenarioRunCount);
-            logFile = createFile( FileLocations.REPORTS_DIRECTORY
+            File logFile = createFile( FileLocations.REPORTS_DIRECTORY
```
Comment (SrinivasanTarget) [inline src/main/java/com/appium/manager/AppiumDevice.java:92]: Good Catch

**Review by SrinivasanTarget — APPROVED** [review]

**Comment by anandbagmar** [comment]
@saikrishna321  @SrinivasanTarget 

The tests are running correctly on local. See attachment for test output. Not sure why the locator is not found for 1 of the tests when running in CI.

[androidCukes.txt](https://github.com/AppiumTestDistribution/AppiumTestDistribution/files/5823739/androidCukes.txt)

**Pull request merged**